# Before you start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
import pandas as pd
import numpy as np


# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents. 

In [2]:
data = pd.read_csv("Current_Employee_Names__Salaries__and_Position_Titles.csv")

Examine the `salaries` dataset using the `head` function below.

In [4]:
display(data.shape)
display(data.head(60))


(33702, 8)

Name                                     Job Titles  \
0     ADRIANO,  RACQUEL ANNE                                      PARAMEDIC   
1          AFFANEH,  MAHIR A                                 POLICE OFFICER   
2          AFOENYI,  PHYLLIS                      TITLE V PROGRAM TRAINEE I   
3            AGAR,  BULENT B                            DEPUTY COMMISSIONER   
4            AGEE,  LAWRENCE                                 POLICE OFFICER   
5          AGNEW,  ANTONIO E                             MOTOR TRUCK DRIVER   
6          AGNEW,  MELISSA M                                       SERGEANT   
7          AGOSTA,  JOSEPH J         POLICE OFFICER (ASSIGNED AS DETECTIVE)   
8    AGOSTINELLI,  MICHAEL A        FIREFIGHTER (PER ARBITRATORS AWARD)-EMT   
9          AGOSTINI,  LUIS R                                       ASST DIR   
10         AGOSTO,  HECTOR I          POLICE OFFICER/EXPLSV DETECT K9 HNDLR   
11          AGREDANO,  ELIAS                                 POLICE OFFICER   
12          AGREDANO,  MARIO                                BRIDGE OPERATOR   
13           AGRON,  EDUARDO                                       SERGEANT   
14             AGRON,  EDWIN                                 POLICE OFFICER   
15        AGSALUD,  FERNAN S                             SANITATION LABORER   
16    AGSALUD JR,  JUANITO S                             SANITATION LABORER   
17    AGUADA,  GILBERT GLENN                                 POLICE OFFICER   
18              AGUAS,  ALEX                                 POLICE OFFICER   
19          AGUAYO,  BRAULIO                                FIREFIGHTER-EMT   
20           AGUAYO,  LUIS A                                FIREFIGHTER-EMT   
21           AGUAYO,  LUIS M                           LABORER - APPRENTICE   
22          AGUAYO,  MARIA G                       ACCOUNTING TECHNICIAN II   
23        AGUIAR,  WILLIAM M                          ADMINISTRATIVE DEPUTY   
24          AGUILA,  JUANITA                             MOTOR TRUCK DRIVER   
25          AGUILAR,  ANDRIA                                ALDERMANIC AIDE   
26          AGUILAR,  EFRAIN                             SANITATION LABORER   
27         AARON,  JEFFERY M                                       SERGEANT   
28            AARON,  KARINA         POLICE OFFICER (ASSIGNED AS DETECTIVE)   
29       AARON,  KIMBERLEI R                       CHIEF CONTRACT EXPEDITER   
30       ABAD JR,  VICENTE M                              CIVIL ENGINEER IV   
31        ABARCA,  FRANCES J                                 POLICE OFFICER   
32         ABASCAL,  REECE E                    TRAFFIC CONTROL AIDE-HOURLY   
33     ABBATACOLA,  ROBERT J                            ELECTRICAL MECHANIC   
34     ABBATEMARCO,  JAMES J                              FIRE ENGINEER-EMT   
35          ABBATE,  TERRY M                                 POLICE OFFICER   
36          ABBOTT,  BETTY L                             FOSTER GRANDPARENT   
37         ABBOTT,  CARMELLA                                 POLICE OFFICER   
38        ABDALLAH,  MARAM M                                      PARAMEDIC   
39           ABDALLAH,  ZAID                                 POLICE OFFICER   
40     ABDELHADI,  ABDALMAHD                                 POLICE OFFICER   
41       ABDELLATIF,  AREF R  FIREFIGHTER (PER ARBITRATORS AWARD)-PARAMEDIC   
42       ABDELLATIF,  HASSAN                                 POLICE OFFICER   
43        ABDELMAJEID,  AZIZ                                       SERGEANT   
44       ABDOLLAHZADEH,  ALI                          FIREFIGHTER/PARAMEDIC   
45  ABDUL-KARIM,  MUHAMMAD A                      ENGINEERING TECHNICIAN VI   
46       ABDULLAH,  DANIEL N                                FIREFIGHTER-EMT   
47         ABDULLAH,  RASHAD               ELECTRICAL MECHANIC (AUTOMOTIVE)   
48        ABDULLA,  MARYEM E                     ASST CORPORATION COUNSEL I   
49    ABDUL-SHAKUR,  TAHIR J                        POOL MOTOR TRUCK DRIVE

We see from looking at the `head` function that there is quite a bit of missing data. Let's examine how much missing data is in each column. Produce this output in the cell below

In [13]:
for column in data:
    print(column, data[column].isna().sum())

Name 0
Job Titles 0
Department 0
Full or Part-Time 0
Salary or Hourly 0
Typical Hours 25528
Annual Salary 8174
Hourly Rate 25528


Let's also look at the count of hourly vs. salaried employees. Write the code in the cell below

In [14]:
data['Salary or Hourly'].value_counts()

Salary    25528
Hourly     8174
Name: Salary or Hourly, dtype: int64

What this information indicates is that the table contains information about two types of employees - salaried and hourly. Some columns apply only to one type of employee while other columns only apply to another kind. This is why there are so many missing values. Therefore, we will not do anything to handle the missing values.

There are different departments in the city. List all departments and the count of employees in each department.

In [16]:
data['Department'].value_counts()

POLICE                 13848
FIRE                    4627
STREETS & SAN           2175
WATER MGMNT             1914
AVIATION                1856
OEMC                    1737
TRANSPORTN              1113
PUBLIC LIBRARY          1088
DAIS                    1033
FAMILY & SUPPORT         656
FINANCE                  544
HEALTH                   467
LAW                      398
CITY COUNCIL             362
BUILDINGS                258
PUBLIC SAFETY ADMIN      244
BUSINESS AFFAIRS         167
HOUSING & ECON DEV       154
COPA                     123
BOARD OF ELECTION        111
MAYOR'S OFFICE           104
INSPECTOR GEN             95
CITY CLERK                88
PROCUREMENT               84
HUMAN RESOURCES           75
ANIMAL CONTRL             75
CULTURAL AFFAIRS          74
HOUSING                   66
BUDGET & MGMT             44
ADMIN HEARNG              38
DISABILITIES              28
TREASURER                 28
HUMAN RELATIONS           17
BOARD OF ETHICS            8
POLICE BOARD  

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [18]:
from scipy.stats import ttest_1samp

In [21]:
# Since I only want to know about the hourly workers, I'll extract those:

hourlydata = data[data['Salary or Hourly'] == 'Hourly']

# And then get just the hourly rates
hourly = hourlydata['Hourly Rate']

#### So the null hypothesis is H₀: μ=30, and H₁: μ≠30. (In this last bit, the "≠" makes it "two sided" - i.e. it could be either higher or lower than 30).

In [27]:
stat, pval = ttest_1samp(hourly, 30)

In [26]:
print(pval)

3.825475847673794e-202


#### This is an absolutely tiny p value - much lower than 0.05. So that means we reject the null hypothesis - the hourly wage is 'significantly different' from 30/h in the sense that the real mean of any sample is extremely unlikely to be 30/hr. 

#### Of course you could quibble about what constitutes 'significantly' different - but this is something already.
##  
##  

We are also curious about salaries in the police force. The chief of police in Chicago claimed in a press briefing that salaries this year are higher than last year's mean of $86000/year a year for all salaried employees. Test this one sided hypothesis using a 95% confidence interval.

Hint: A one tailed test has a p-value that is half of the two tailed p-value. If our hypothesis is greater than, then to reject, the test statistic must also be positive.

In [39]:
# First I'll extract the relevant data. Let me check if any people in the police are paid by the hour:

len(data[(data['Department'] == 'POLICE') & (data['Salary or Hourly'] == 'Hourly')])

24

In [45]:
# Only 24 - I'll ignore those to make life easy. 

policedata = data[(data['Department'] == 'POLICE') & (data['Salary or Hourly'] != 'Hourly')]
police = policedata['Annual Salary']

#### The null hypothesis here is H0: mean annual salary (for Department == Police) = 86000
#### H1 is: mean annual salary > 86000

In [47]:
stat, pval = ttest_1samp(police, 86000)

print(pval /2) #dividing the p value for the two-tailed test by two - that gives me the p value for the one-tailed test.

6.119566952121482e-76


In [ ]:
# This is, once again, an absolutely tiny value - virtually no chance that the mean of any sampling would be 86000.

In [49]:
# And this checks out with the actual mean of the actual data:
np.mean(police)

88834.11892361111

Using the `crosstab` function, find the department that has the most hourly workers. 

In [51]:
pd.crosstab(data['Department'], data['Salary or Hourly'])

Salary or Hourly     Hourly  Salary
Department                         
ADMIN HEARNG              0      38
ANIMAL CONTRL            12      63
AVIATION               1214     642
BOARD OF ELECTION         0     111
BOARD OF ETHICS           0       8
BUDGET & MGMT             2      42
BUILDINGS                 0     258
BUSINESS AFFAIRS          7     160
CITY CLERK                0      88
CITY COUNCIL             74     288
COPA                      0     123
CULTURAL AFFAIRS         10      64
DAIS                    745     288
DISABILITIES              1      27
FAMILY & SUPPORT        357     299
FINANCE                  45     499
FIRE                      0    4627
HEALTH                    2     465
HOUSING                   0      66
HOUSING & ECON DEV        3     151
HUMAN RELATIONS           0      17
HUMAN RESOURCES           0      75
INSPECTOR GEN             0      95
LAW                      48     350
LICENSE APPL COMM         0       1
MAYOR'S OFFICE           17      87
OEMC                   1077     660
POLICE                   24   13824
POLICE BOARD              0       2
PROCUREMENT               2      82
PUBLIC LIBRARY          356     732
PUBLIC SAFETY ADMIN      48     196
STREETS & SAN          1846     329
TRANSPORTN              726     387
TREASURER                 0      28
WATER MGMNT            1558     356

#### I don't love the crosstab function, since it does not allow me to order the results. So from reading the result, I figured out that the highest value for 'Hourly' is in Streets & San\[itation\], namely 1846.

The workers from the department with the most hourly workers have complained that their hourly wage is less than $35/hour. Using a one sample t-test, test this one-sided hypothesis at the 95% confidence level.

In [64]:
# I'm cleaning out the NaNs and selecting only the 'Hourly Rate' row right away

s_and_s = data[(data['Department'] == 'STREETS & SAN') & (data['Hourly Rate'].notnull())]['Hourly Rate']


s_and_s

15       40.78
16       38.52
26       38.52
49       29.96
65       37.45
         ...  
33623    38.01
33624    37.45
33667    37.45
33668    38.52
33676    25.95
Name: Hourly Rate, Length: 1846, dtype: float64

#### Spelling it all out again: 
- H0: the mean hourly rate = 35/hr
- H1: the mean hourly rate > 35/hr (or <, but the p-value will be the same on either end)


In [67]:
stat, pval = ttest_1samp(s_and_s, 35)

In [69]:
pval/2

3.2902372218867895e-05

Again, the pval is smaller than 0.05 - though this time, not as microscopically small as before.

# Challenge 3: To practice - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [71]:
from scipy.stats import t
from scipy.stats import sem

Now construct the 95% confidence interval for all salaried employeed in the police in the cell below.

#### I already had the salaried police in the variable 'police' so I'll keep using that.

In [72]:
standard_error = sem(police)

In [74]:
confidence_interval = t.interval(0.95, len(police)-1,np.mean(police),standard_error) 

In [75]:
print(confidence_interval)

(88534.1182885883, 89134.11955863392)


# Bonus Challenge - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [76]:
# I'll skip the bonus for today. Cheers!